diff --git a/virt/kvm/kvm_main.c b/virt/kvm/kvm_main.c
index 71d4ab0..39268bd 100644
--- a/virt/kvm/kvm_main.c
+++ b/virt/kvm/kvm_main.c
@@ -3194,11 +3194,13 @@ static int kvm_dev_ioctl_create_vm(unsigned long type)
 		return PTR_ERR(file);
 	}
 
+#ifndef CONFIG_GRKERNSEC_SYSFS_RESTRICT
 	if (kvm_create_vm_debugfs(kvm, r) < 0) {
 		put_unused_fd(r);
 		fput(file);
 		return -ENOMEM;
 	}
+#endif
 
 	fd_install(r, file);
 	return r;
