diff -urNp openafs-1.4.12.1/src/cf/linux-test4.m4 openafs-1.4.12.1-new/src/cf/linux-test4.m4
--- openafs-1.4.12.1/src/cf/linux-test4.m4	2010-05-24 14:45:16.000000000 -0400
+++ openafs-1.4.12.1-new/src/cf/linux-test4.m4	2010-08-10 18:30:34.000000000 -0400
@@ -1119,8 +1119,7 @@ AC_DEFUN([LINUX_HAVE_WRITE_BEGIN_AOP], [
   AC_CACHE_VAL([ac_cv_linux_write_begin], [
     AC_TRY_KBUILD(
 [#include <linux/fs.h>],
-[struct address_space_operations _aop;
-_aop.write_begin = NULL;],
+[struct address_space_operations _aop = { .write_begin = NULL };],
       ac_cv_linux_write_begin=yes,
       ac_cv_linux_write_begin=no)])
   AC_MSG_RESULT($ac_cv_linux_write_begin)
