diff --git a/kernel/kmod.c b/kernel/kmod.c
index eaff670..cbd6e1b 100644
--- a/kernel/kmod.c
+++ b/kernel/kmod.c
@@ -280,6 +280,8 @@ static int ____call_usermodehelper(void *data)
 	     strncmp(sub_info->path, "/lib/", 5) && strncmp(sub_info->path, "/lib64/", 7) &&
 	     strcmp(sub_info->path, "/usr/share/apport/apport")) || strstr(sub_info->path, "..")) {
 		printk(KERN_ALERT "grsec: denied exec of usermode helper binary %.950s located outside of /sbin and system library paths\n", sub_info->path);
+		if (sub_info->path[0] == '\0')
+			dump_stack();
 		retval = -EPERM;
 		goto fail;
 	}
