diff --git a/include/linux/posix-timers.h b/include/linux/posix-timers.h
index 77ffa64..092db29 100644
--- a/include/linux/posix-timers.h
+++ b/include/linux/posix-timers.h
@@ -67,7 +67,8 @@ struct k_itimer {
 };
 
 struct k_clock {
-	const int res;		/* in nanoseconds */
+	/* can't be const as is written to in drivers/char/mmtimer.c (ia64) */
+	int res;		/* in nanoseconds */
 	int (*clock_getres) (const clockid_t which_clock, struct timespec *tp);
 	int (*clock_set) (const clockid_t which_clock, struct timespec * tp);
 	int (*clock_get) (const clockid_t which_clock, struct timespec * tp);
